src/fparser/two/Fortran2008.py

```diff
@@ -96,6 +101,241 @@ class Program_Unit(Program_Unit_2003):  # R202
+                  | ALLOCATABLE
+                  | ASYNCHRONOUS
+                  | CODIMENSION <lbracket> <coarray-spec> <rbracket>
+                  | CONTIGUOUS
+                  | DIMENSION ( <array-spec> )
+                  | EXTERNAL
+                  | INTENT ( <intent-spec> )
+                  | INTRINSIC
+                  | <language-binding-spec>
+                  | OPTIONAL
+                  | PARAMETER
+                  | POINTER
+                  | PROTECTED
+                  | SAVE
+                  | TARGET
+                  | VALUE
+                  | VOLATILE
+    """
+    subclass_names = ['Access_Spec', 'Language_Binding_Spec',
+                      'Dimension_Attr_Spec', 'Intent_Attr_Spec',
+                      'Codimension_Attr_Spec']
+    use_names = []
+
+    @staticmethod
+    def match(string):
+        '''Implements the matching for a types attributes.
+
+        :param str string: the string to match as attribute.
+        :return None if there is no match, otherwise the matched object.
+        :rtype: None, Attr_Spec, Access_Spec, Language_Binding_Spec, \
+            Dimension_Attr_Spec, Intent_Attr_Spec, Codimension_Attr_Spec
+        '''
+        return STRINGBase.match(pattern.abs_attr_spec_f08, string)
+
+
+class Coarray_Spec(Base):  # R509
+    """
+:F08R:`509`
+    <coarray-spec> = <deferred-coshape-spec-list>
+                      | <explicit-coshape-spec-list>
+    """
+    subclass_names = ['Explicit_Coshape_Spec',
+                      'Deferred_Coshape_Spec_List']
+
+
+class Deferred_Coshape_Spec(SeparatorBase):  # R510
+    """
+:F08R:`510`
+    <deferred-coshape-spec> = :
+    """
+    subclass_names = []
+
+    @staticmethod
+    def match(string):
+        '''Implements the matching for deferred coarray shape specification.
+
+        :param str string: the string to match as deferred shape.
+        :return None if there is no match, otherwise the matched object.
+        :rtype: None, Deferred_Coshape_Spec
+        '''
+        if string == ':':
+            return None, None
+        return
+
+
+class Explicit_Coshape_Spec(SeparatorBase):  # R511
+    """
+:F08R:`511`
+    <explicit-coshape-spec> = [ <coshape-spec-list> , ]
+        [ <lower-cobound> : ] *
+    """
+    subclass_names = []
+    use_names = ['Coshape_Spec_List', 'Lower_Cobound']
+
+    @staticmethod
+    def match(string):
+        '''Implements the matching for explicit coarray shape specification.
+
+        :param str string: the string to match as deferred shape.
+        :return None if there is no match, otherwise the matched object.
+        :rtype: None, Explicit_Coshape_Spec, Coshape_Spec_List
+        '''
+        if not string.endswith('*'):
+            return
+        line = string[:-1].rstrip()
+        if not line:
+            return None, None
+        if line.endswith(':'):
+            line, repmap = string_replace_map(line[:-1].rstrip())
+            i = line.rfind(',')
```

Comment:
Please check all changes with pycodestyle and pylint as I think `i` will cause them to complain.